<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Built-in-Data-Types-and-casting" data-toc-modified-id="Built-in-Data-Types-and-casting-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Built-in Data Types and casting</a></span><ul class="toc-item"><li><span><a href="#Sequence-Type" data-toc-modified-id="Sequence-Type-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Sequence Type</a></span></li><li><span><a href="#Getting-the-Data-Type" data-toc-modified-id="Getting-the-Data-Type-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Getting the Data Type</a></span></li><li><span><a href="#Setting-the-Data-Type" data-toc-modified-id="Setting-the-Data-Type-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Setting the Data Type</a></span></li><li><span><a href="#Setting-the-Specific-Data-Type" data-toc-modified-id="Setting-the-Specific-Data-Type-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Setting the Specific Data Type</a></span></li></ul></li><li><span><a href="#Python-Numbers" data-toc-modified-id="Python-Numbers-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Python Numbers</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#int" data-toc-modified-id="int-2.0.1"><span class="toc-item-num">2.0.1&nbsp;&nbsp;</span>int</a></span></li><li><span><a href="#Float" data-toc-modified-id="Float-2.0.2"><span class="toc-item-num">2.0.2&nbsp;&nbsp;</span>Float</a></span></li><li><span><a href="#Complex" data-toc-modified-id="Complex-2.0.3"><span class="toc-item-num">2.0.3&nbsp;&nbsp;</span>Complex</a></span></li><li><span><a href="#Type-Conversion-within-Numeric-Type" data-toc-modified-id="Type-Conversion-within-Numeric-Type-2.0.4"><span class="toc-item-num">2.0.4&nbsp;&nbsp;</span>Type Conversion within Numeric Type</a></span></li><li><span><a href="#Random-Number" data-toc-modified-id="Random-Number-2.0.5"><span class="toc-item-num">2.0.5&nbsp;&nbsp;</span>Random Number</a></span></li><li><span><a href="#Python-Casting" data-toc-modified-id="Python-Casting-2.0.6"><span class="toc-item-num">2.0.6&nbsp;&nbsp;</span>Python Casting</a></span></li></ul></li></ul></li><li><span><a href="#Python-Strings" data-toc-modified-id="Python-Strings-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Python Strings</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Assign-String-to-a-Variable" data-toc-modified-id="Assign-String-to-a-Variable-3.0.1"><span class="toc-item-num">3.0.1&nbsp;&nbsp;</span>Assign String to a Variable</a></span></li><li><span><a href="#Multiline-Strings" data-toc-modified-id="Multiline-Strings-3.0.2"><span class="toc-item-num">3.0.2&nbsp;&nbsp;</span>Multiline Strings</a></span></li><li><span><a href="#Strings-are-Arrays" data-toc-modified-id="Strings-are-Arrays-3.0.3"><span class="toc-item-num">3.0.3&nbsp;&nbsp;</span>Strings are Arrays</a></span></li><li><span><a href="#Looping-Through-a-String" data-toc-modified-id="Looping-Through-a-String-3.0.4"><span class="toc-item-num">3.0.4&nbsp;&nbsp;</span>Looping Through a String</a></span></li><li><span><a href="#len()" data-toc-modified-id="len()-3.0.5"><span class="toc-item-num">3.0.5&nbsp;&nbsp;</span>len()</a></span></li><li><span><a href="#Check-String-(in)" data-toc-modified-id="Check-String-(in)-3.0.6"><span class="toc-item-num">3.0.6&nbsp;&nbsp;</span>Check String (in)</a></span></li><li><span><a href="#Check-if-NOT-(not-in)" data-toc-modified-id="Check-if-NOT-(not-in)-3.0.7"><span class="toc-item-num">3.0.7&nbsp;&nbsp;</span>Check if NOT (not in)</a></span></li><li><span><a href="#Slicing-Strings(SubString/indexing-)" data-toc-modified-id="Slicing-Strings(SubString/indexing-)-3.0.8"><span class="toc-item-num">3.0.8&nbsp;&nbsp;</span>Slicing Strings(SubString/indexing )</a></span></li><li><span><a href="#Setp-size-in-slicing" data-toc-modified-id="Setp-size-in-slicing-3.0.9"><span class="toc-item-num">3.0.9&nbsp;&nbsp;</span>Setp size in slicing</a></span></li><li><span><a href="#String---Upper()" data-toc-modified-id="String---Upper()-3.0.10"><span class="toc-item-num">3.0.10&nbsp;&nbsp;</span>String - Upper()</a></span></li><li><span><a href="#String---lower-()" data-toc-modified-id="String---lower-()-3.0.11"><span class="toc-item-num">3.0.11&nbsp;&nbsp;</span>String - lower ()</a></span></li><li><span><a href="#String-replace()" data-toc-modified-id="String-replace()-3.0.12"><span class="toc-item-num">3.0.12&nbsp;&nbsp;</span>String replace()</a></span></li><li><span><a href="#String-split()" data-toc-modified-id="String-split()-3.0.13"><span class="toc-item-num">3.0.13&nbsp;&nbsp;</span>String split()</a></span></li><li><span><a href="#String-strip()" data-toc-modified-id="String-strip()-3.0.14"><span class="toc-item-num">3.0.14&nbsp;&nbsp;</span>String strip()</a></span></li><li><span><a href="#String-Concatenation" data-toc-modified-id="String-Concatenation-3.0.15"><span class="toc-item-num">3.0.15&nbsp;&nbsp;</span>String Concatenation</a></span></li></ul></li></ul></li><li><span><a href="#Format---Strings-(format())" data-toc-modified-id="Format---Strings-(format())-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Format - Strings (format())</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#String-Formatting" data-toc-modified-id="String-Formatting-4.0.1"><span class="toc-item-num">4.0.1&nbsp;&nbsp;</span>String Formatting</a></span></li></ul></li><li><span><a href="#Formatting-with-placeholders-(%)-(%s,-%r)" data-toc-modified-id="Formatting-with-placeholders-(%)-(%s,-%r)-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Formatting with placeholders (%) (%s, %r)</a></span><ul class="toc-item"><li><span><a href="#Format-conversion-methods." data-toc-modified-id="Format-conversion-methods.-4.1.1"><span class="toc-item-num">4.1.1&nbsp;&nbsp;</span>Format conversion methods.</a></span></li><li><span><a href="#repr()" data-toc-modified-id="repr()-4.1.2"><span class="toc-item-num">4.1.2&nbsp;&nbsp;</span>repr()</a></span></li><li><span><a href="#str()" data-toc-modified-id="str()-4.1.3"><span class="toc-item-num">4.1.3&nbsp;&nbsp;</span>str()</a></span></li><li><span><a href="#Padding-and-Precision-of-Floating-Point-Numbers" data-toc-modified-id="Padding-and-Precision-of-Floating-Point-Numbers-4.1.4"><span class="toc-item-num">4.1.4&nbsp;&nbsp;</span>Padding and Precision of Floating Point Numbers</a></span></li><li><span><a href="#Multiple-Formatting-with-placeholders-(%)" data-toc-modified-id="Multiple-Formatting-with-placeholders-(%)-4.1.5"><span class="toc-item-num">4.1.5&nbsp;&nbsp;</span>Multiple Formatting with placeholders (%)</a></span></li></ul></li><li><span><a href="#Formatting-with-the-.format()-method" data-toc-modified-id="Formatting-with-the-.format()-method-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Formatting with the <code>.format()</code> method</a></span><ul class="toc-item"><li><span><a href="#String-Formatting-using--format()" data-toc-modified-id="String-Formatting-using--format()-4.2.1"><span class="toc-item-num">4.2.1&nbsp;&nbsp;</span>String Formatting using  format()</a></span></li><li><span><a href="#The-.format()-method-has-several-advantages-over-the-%s-placeholder-method:" data-toc-modified-id="The-.format()-method-has-several-advantages-over-the-%s-placeholder-method:-4.2.2"><span class="toc-item-num">4.2.2&nbsp;&nbsp;</span>The .format() method has several advantages over the %s placeholder method:</a></span><ul class="toc-item"><li><span><a href="#1.-Inserted-objects-can-be-called-by-index-position:" data-toc-modified-id="1.-Inserted-objects-can-be-called-by-index-position:-4.2.2.1"><span class="toc-item-num">4.2.2.1&nbsp;&nbsp;</span>1. Inserted objects can be called by index position:</a></span></li><li><span><a href="#2.-Inserted-objects-can-be-assigned-keywords:" data-toc-modified-id="2.-Inserted-objects-can-be-assigned-keywords:-4.2.2.2"><span class="toc-item-num">4.2.2.2&nbsp;&nbsp;</span>2. Inserted objects can be assigned keywords:</a></span></li><li><span><a href="#3.-Inserted-objects-can-be-reused,-avoiding-duplication:" data-toc-modified-id="3.-Inserted-objects-can-be-reused,-avoiding-duplication:-4.2.2.3"><span class="toc-item-num">4.2.2.3&nbsp;&nbsp;</span>3. Inserted objects can be reused, avoiding duplication:</a></span></li></ul></li><li><span><a href="#Alignment,-padding-and-precision-with-.format()" data-toc-modified-id="Alignment,-padding-and-precision-with-.format()-4.2.3"><span class="toc-item-num">4.2.3&nbsp;&nbsp;</span>Alignment, padding and precision with <code>.format()</code></a></span></li></ul></li><li><span><a href="#Formatted-String-Literals-(f-strings)" data-toc-modified-id="Formatted-String-Literals-(f-strings)-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Formatted String Literals (f-strings)</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Float-formatting-follows-&quot;result:-{value:{width}.{precision}}&quot;" data-toc-modified-id="Float-formatting-follows-&quot;result:-{value:{width}.{precision}}&quot;-4.3.0.1"><span class="toc-item-num">4.3.0.1&nbsp;&nbsp;</span>Float formatting follows <code>"result: {value:{width}.{precision}}"</code></a></span></li></ul></li></ul></li><li><span><a href="#Python---Escape-Characters" data-toc-modified-id="Python---Escape-Characters-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Python - Escape Characters</a></span></li></ul></li><li><span><a href="#Python---String-Methods" data-toc-modified-id="Python---String-Methods-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Python - String Methods</a></span><ul class="toc-item"><li><span><a href="#capitalize()" data-toc-modified-id="capitalize()-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>capitalize()</a></span></li><li><span><a href="#casefold()" data-toc-modified-id="casefold()-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>casefold()</a></span></li><li><span><a href="#center()" data-toc-modified-id="center()-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>center()</a></span></li><li><span><a href="#count()" data-toc-modified-id="count()-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>count()</a></span></li><li><span><a href="#lower()" data-toc-modified-id="lower()-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>lower()</a></span></li><li><span><a href="#upper()" data-toc-modified-id="upper()-5.6"><span class="toc-item-num">5.6&nbsp;&nbsp;</span>upper()</a></span></li></ul></li></ul></div>

# Built-in Data Types and casting

https://www.geeksforgeeks.org/python-data-types/

https://www.w3schools.com/python/python_datatypes.asp

In programming, data type is an important concept.

Variables can store data of different types, and different types can do different things.

Python has the following data types built-in by default, in these categories:

`Text Type:	str
Numeric Types:	int, float, complex
Sequence Types:	list, tuple, range
Mapping Type:	dict
Set Types:	set, frozenset
Boolean Type:	bool
Binary Types:	bytes, bytearray, memoryview`

__Python Data Types__

https://www.tutorialsteacher.com/python/python-data-types

Data types are the classification or categorization of data items. Python supports the following built-in data types.

* Scalar Types
* Sequence Type
* Mapping Type
* Set Types


## Sequence Type

https://www.tutorialsteacher.com/python/python-data-types

__In Python, sequence is the ordered collection of similar or different data types. Sequences allows to store multiple values in an organized and efficient fashion. There are several sequence types in Python__

Python has the following built-in sequence data types:

__String__: A string value is a collection of one or more characters put in single, double or triple quotes.

__List__: A list object is an ordered collection of one or more data items, not necessarily of the same type, put in square brackets.

__Tuple__ :  A Tuple object is an ordered collection of one or more data items, not necessarily of the same type, put in parentheses.

__range__ :  range() is a Python __Data type__  and a Function as well.



__What is a sequence data type in Python?__
https://www.tutorialspoint.com/What-is-a-sequence-data-type-in-Python

_Sequences allow you to store multiple values in an organized and efficient fashion._ There are several sequence types: strings, Unicode strings, lists, tuples, bytearrays, and range objects. Dictionaries and sets are containers for non-sequential data.

From the official Python Docs −

Strings are immutable sequences of Unicode code points.

Lists are mutable sequences, typically used to store collections of homogeneous items.

Tuples are immutable sequences, typically used to store collections of heterogeneous data (such as the 2-tuples produced by the enumerate() built-in).

Bytearray objects are mutable, they support the mutable sequence operations in addition to the common bytes and bytearray operations

The range type represents an immutable sequence of numbers and is commonly used for looping a specific number of times in for loops.

## Getting the Data Type
You can get the data type of any object by using the type() function:

In [1]:
# Print the data type of the variable x:
x = 5
print(type(x))

<class 'int'>


## Setting the Data Type

In Python, the data type is set when you assign a value to a variable:

In [2]:
x = "Hello World"
print(x, type(x))
x = 20.5
print(x, type(x))
x = 1j
print(x, type(x))
x = ["apple", "banana", "cherry"]
print(x, type(x))
x = ("apple", "banana", "cherry")
print(x, type(x))
x = range(6)
print(x, type(x))
x = {"name": "John", "age": 36}
print(x, type(x))
x = {"apple", "banana", "cherry"}
print(x, type(x))
x = frozenset({"apple", "banana", "cherry"})
print(x, type(x))
x = True
print(x, type(x))
x = b"Hello"
print(x, type(x))
x = bytearray(5)
print(x, type(x))
x = memoryview(bytes(5))
print(x, type(x))

Hello World <class 'str'>
20.5 <class 'float'>
1j <class 'complex'>
['apple', 'banana', 'cherry'] <class 'list'>
('apple', 'banana', 'cherry') <class 'tuple'>
range(0, 6) <class 'range'>
{'name': 'John', 'age': 36} <class 'dict'>
{'cherry', 'apple', 'banana'} <class 'set'>
frozenset({'cherry', 'apple', 'banana'}) <class 'frozenset'>
True <class 'bool'>
b'Hello' <class 'bytes'>
bytearray(b'\x00\x00\x00\x00\x00') <class 'bytearray'>
<memory at 0x000001F06F279280> <class 'memoryview'>


## Setting the Specific Data Type

If you want to specify the data type, you can use the following constructor functions:

In [3]:
x = str("Hello World")
print(x, type(x))

x = int(20)
print(x, type(x))

x = float(20.5)
print(x, type(x))

x = complex(1j)
print(x, type(x))

x = list(("apple", "banana", "cherry"))
print(x, type(x))

x = tuple(("apple", "banana", "cherry"))
print(x, type(x))

x = range(6)
print(x, type(x))

x = dict(name="John", age=36)
print(x, type(x))

x = set(("apple", "banana", "cherry"))
print(x, type(x))

x = frozenset(("apple", "banana", "cherry"))
print(x, type(x))

x = bool(5)
print(x, type(x))

x = bytes(5)
print(x, type(x))

x = bytearray(5)
print(x, type(x))

x = memoryview(bytes(5))
print(x, type(x))


Hello World <class 'str'>
20 <class 'int'>
20.5 <class 'float'>
1j <class 'complex'>
['apple', 'banana', 'cherry'] <class 'list'>
('apple', 'banana', 'cherry') <class 'tuple'>
range(0, 6) <class 'range'>
{'name': 'John', 'age': 36} <class 'dict'>
{'cherry', 'apple', 'banana'} <class 'set'>
frozenset({'cherry', 'apple', 'banana'}) <class 'frozenset'>
True <class 'bool'>
b'\x00\x00\x00\x00\x00' <class 'bytes'>
bytearray(b'\x00\x00\x00\x00\x00') <class 'bytearray'>
<memory at 0x000001F06F279280> <class 'memoryview'>


# Python Numbers

There are three numeric types in Python:

int

float

complex

Variables of numeric types are created when you assign a value to them:

In [4]:
# To verify the type of any object in Python, use the type() function:
    
x = 1    # int
print(x, type(x))

y = 2.8  # float
print(y, type(y))

z = 1j   # complex
print(z, type(z))

1 <class 'int'>
2.8 <class 'float'>
1j <class 'complex'>


### int 

* Int, or integer, is a whole number, positive or negative, without decimals, of unlimited length.

In [5]:
x = 1
y = 35656222554887711
z = -3255522

print(x, type(x))
print(y, type(y))
print(z, type(z))

1 <class 'int'>
35656222554887711 <class 'int'>
-3255522 <class 'int'>


In [6]:
x = 1.10
y = 1.0
z = -35.59
print(x, type(x))
print(y, type(y))
print(z, type(z))

1.1 <class 'float'>
1.0 <class 'float'>
-35.59 <class 'float'>


### Float
* Float, or "floating point number" is a number, positive or negative, containing one or more decimals.

In [7]:
x = 1.10
y = 1.0
z = -35.59

print(x, type(x))
print(y, type(y))
print(z, type(z))

1.1 <class 'float'>
1.0 <class 'float'>
-35.59 <class 'float'>


__Imp point for float Data Type__
* Float can also be scientific numbers with an "e" to indicate the power of 10.0
* Remember by default in Python "E" or "e" means 10.0. 
* Captial "E" and small "e" both provides same results
* To use Euler's number ("e" which value is 2.718281828459045) you need to use Either Numpy or Math Library

In [8]:
1e+1

10.0

In [9]:
1E1

10.0

In [10]:
y1 = 35e1 # (35 * 10.0 ^1 )
y1

350.0

In [11]:
x = 35e3
y = 12E4
z = -87.7e100

print(x, type(x))
print(y, type(y))
print(z, type(z))

35000.0 <class 'float'>
120000.0 <class 'float'>
-8.77e+101 <class 'float'>


In [12]:
print("1e1:",1e1)
print(1e+1)
print(1e+2)
print(1e+3)
print(1e-1)
print(1e-2)
print(1e-3)
print(1.7e+10)
print(1.7e-4)
print(3.4e+10)

1e1: 10.0
10.0
100.0
1000.0
0.1
0.01
0.001
17000000000.0
0.00017
34000000000.0


In [13]:
q = -87.7e10
q

-877000000000.0

In [14]:
x = -87.7e100
print(x, type(x))

-8.77e+101 <class 'float'>


result is (8.77 * 10.0 power 101)

### Complex
* Complex numbers are written with a "j" as the imaginary part:

In [15]:
x = 3+5j
y = 5j
z = -5j
print(x, type(x))
print(y, type(y))
print(z, type(z))

(3+5j) <class 'complex'>
5j <class 'complex'>
(-0-5j) <class 'complex'>


### Type Conversion within Numeric Type
You can convert from one type to another with the int(), float(), and complex() methods:

`Note: You cannot convert complex numbers into another number type.`

In [16]:
x = 1    # int
y = 2.8  # float
z = 1j   # complex

print(x, type(x))
print(y, type(y))
print(z, type(z))

print("\n *******converted data")
# convert from int to float:
a = float(x)
# convert from float to int:
b = int(y)
# convert from int to complex:
c = complex(x)

print(a, type(a))
print(b, type(b))
print(c, type(c))

1 <class 'int'>
2.8 <class 'float'>
1j <class 'complex'>

 *******converted data
1.0 <class 'float'>
2 <class 'int'>
(1+0j) <class 'complex'>


### Random Number
Python does not have a random() function to make a random number, but Python has a built-in module called `random`
that can be used to make random numbers

* In Random Module Reference(https://www.w3schools.com/python/module_random.asp) you will learn more about the Random module.

In [17]:
# Import the random module, and display a random number between 1 and 9:

# Signature: random.randrange(start, stop=None, step=1, _int=<class 'int'>)
# Docstring:
# Choose a random item from range(start, stop[, step]).
# 
# This fixes the problem with randint() which includes the
# endpoint; in Python this is usually not what you want.
# File:      c:\users\sachin\anaconda3\lib\random.py
# Type:      method

import random

print(random.randrange(1, 10)) #returns a number between 1 (included) and 10 (not included)

8


### Python Casting

__Specify a Variable Type__

There may be times when you want to specify a type on to a variable. This can be done with casting. Python is an object-orientated language, and as such it uses classes to define data types, including its primitive types.

Casting in python is therefore done using constructor functions:

`int()  `
constructs an integer number from an integer literal, a float literal (by removing all decimals), or a string literal (providing the string represents a whole number)

`float()`
 constructs a float number from an integer literal, a float literal or a string literal (providing the string represents a float or an integer)

`str()` - constructs a string from a wide variety of data types, including strings, integer literals and float literals


In [18]:
x = int(1)
y = int(2.8)
z = int("3")

print(1,    type(1), x, type(x))
print(2.8,  type(2.8), y, type(y))
print("3",  type("3"), z, type(z))

1 <class 'int'> 1 <class 'int'>
2.8 <class 'float'> 2 <class 'int'>
3 <class 'str'> 3 <class 'int'>


In [19]:
x = float(1)     # x will be 1.0
y = float(2.8)   # y will be 2.8
z = float("3")   # z will be 3.0
w = float("4.2")  # w will be 4.2
print(1,    type(1), x, type(x))
print(2.8,  type(2.8), y, type(y))
print("3",  type("3"), z, type(z))
print("4.2", type("4.2"), w, type(w))

1 <class 'int'> 1.0 <class 'float'>
2.8 <class 'float'> 2.8 <class 'float'>
3 <class 'str'> 3.0 <class 'float'>
4.2 <class 'str'> 4.2 <class 'float'>


In [20]:
x = str("s1") # x will be 's1'
y = str(2)    # y will be '2'
z = str(3.0)  # z will be '3.0'

print(1,    type(1), x, type(x))
print(2.8,  type(2.8), y, type(y))
print("3",  type("3"), z, type(z))

1 <class 'int'> s1 <class 'str'>
2.8 <class 'float'> 2 <class 'str'>
3 <class 'str'> 3.0 <class 'str'>


# Python Strings

Learn more about If statements in  Python If...Else chapter (https://www.w3schools.com/python/python_conditions.asp)

Learn more about For Loops in Python For Loops(https://www.w3schools.com/python/python_for_loops.asp) 

Learn more about String Methods with (https://www.w3schools.com/python/python_ref_string.asp)

Learn more about Lists in (https://www.w3schools.com/python/python_lists.asp)



__Strings__

* Strings literals in python are surrounded by either single quotation marks, or double quotation marks. 
* 'hello' is the same as "hello".
* You can display a string literal with the print() function:For example: print("hello").

In [21]:
'Hello' # with single quote

'Hello'

In [22]:
"Hello" # with Double  quote

'Hello'

In [23]:
# Print doesn’t PRINT quotes it just print actual Value without single/Double quotes
print("Hello")
print('Hello')

Hello
Hello


### Assign String to a Variable

Assigning a string to a variable is done with the variable name followed by an equal sign and the string:

In [24]:
a = "Hello"
a

'Hello'

In [25]:
print(a)

Hello


### Multiline Strings

You can assign a multiline string to a variable by using three quotes:

You can use three double quotes:

In [26]:
a = """Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua."""
print(a)

Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.


> Or three single quotes

In [27]:
a = '''Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.'''
print(a)

Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.


`Note: in the result, the line breaks are inserted at the same position as in the code.`  --> ???

### Strings are Arrays
Like many other popular programming languages, strings in Python are arrays of bytes representing unicode characters.

However, Python does not have a character data type, a single character is simply a string with a length of 1.

__Square brackets can be used to access elements of the string.__

Get the character at position 1 (`remember that the first character has the index 0`):

In [28]:
a = "Hello, World!"

In [29]:
print(a[0])

H


In [30]:
print(a[1])

e


In [31]:
print(a[12])

!


In [32]:
print(a[13])

IndexError: string index out of range

In [33]:
len(a) # length is 13 char starting from 0(H) to 12(!)

13

### Looping Through a String

Since strings are arrays, we can loop through the characters in a string, with a for loop.



In [34]:
# Loop through the letters in the word "banana"
for x in "banana":
  print(x)

b
a
n
a
n
a


### len() 

To get the length of a string, use the len() function.

The `len()` function returns the length of a string/object :


In [35]:
a = "Hello, World!"
print(len(a))

13


### Check String (in)

To check if a certain phrase or character is present in a string, we can use the keyword `in`.

Check if "free" is present in the following text:

In [36]:
txt = "The best things in life are free!"
print("free" in txt)

True


* Use it in an if statement

Print only if "free" is present:

In [37]:
txt = "The best things in life are free!"
if "free" in txt:
    print("Yes, 'free' is present.")

Yes, 'free' is present.


### Check if NOT (not in)

To check if a certain phrase or character is NOT present in a string, we can use the keyword `not in`

In [38]:
# Check if "expensive" is NOT present in the following text:
txt = "The best things in life are free!"
print("expensive" not in txt)

# Use it in an if statement:
# print only if "expensive" is NOT present:
txt = "The best things in life are free!"
if "expensive" not in txt:
    print("No, 'expensive' is NOT present.")

True
No, 'expensive' is NOT present.


### Slicing Strings(SubString/indexing )

__Slicing__

You can return a range of characters by using the slice syntax.

Specify the start index and the end index, separated by a colon, to return a part of the string.

`Note: The first character has index 0.`

* String [START:END:STEP SIZE ]  
here 
* START : Start from START index and include START index  and
* END : end at END index -1 index means excluding END index 
* STEP size is : how you want to take steps. Its could be “-“ or “+” number 


In [39]:
b = "Hello, World!"
b

'Hello, World!'

In [40]:
b = "Hello, World!"
print(b[0])
print(b[1])
print(b[2])
print(b[3])
print(b[4])
print(b[5])
print(b[6])
print(b[7])

H
e
l
l
o
,
 
W


In [41]:
x = "Hello, World!"
x[13:]

''

here '' is not a value its means nothing is there  if you see len and type of (x[13:]) then it will be string of length zero 


In [42]:
print(type(x[13:]))
print(len(x[13:]))

<class 'str'>
0


In [43]:
y = " "
y

' '

In [44]:
print(type(y[13:]))
print(len(y[13:]))

<class 'str'>
0


In [45]:
b = "Hello, World!"
b[:5]

'Hello'

In [46]:
b = "Hello, World!"
b[1:9:2]  # start from 1 till index (9-1) 8 and STEp by 2
# 1
# 1+2 = 3
# 3+2 = 5
# 5+2 = 7

'el,W'

In [47]:
b = "0123456789"
b[1:9:2]

'1357'

In [48]:
b = "Hello, World!"
b[0:5]

'Hello'

In [49]:
b = "Hello, World!"
b[2:5]

'llo'

In [50]:
b = "Hello, World!"
b[5]

','

In [51]:
b = "Hello, World!"
b[11:12]  # here we are saying get data from index 11 till (12-1) 11 index means straing from 11 to 11 only 
# which is 'd'

'd'

In [52]:
b = "Hello, World!"
c = b[7:10+1]
c

'Worl'

In [53]:
len(c)

4

__Slice From the Start__
By leaving out the start index, the range will start at the first character:

In [54]:
# Get the characters from the start to position 5 (not included):
b = "Hello, World!"
print(b[:5])

Hello


Note the above slicing. Here we're telling Python to grab everything from index 0 up to 5. It doesn't include the 5th index. You'll notice this a lot in Python, where statements and are usually in the context of __"up to, but not including".__

__Slice To the End__

By leaving out the end index, the range will go to the end:

In [55]:
# Get the characters from position 2, and all the way to the end:
b = "Hello, World!"
print(b[2:])

llo, World!


In [56]:
#Everything. Nothing mentioned hence start from 0 index and end on the last index 
b = "Hello, World!"
print(b[:])

Hello, World!


__Negative Indexing__

Use negative indexes to start the slice from the end of the string:


In [57]:
# Get the characters:
# From: "o" in "World!" (position -5)
# To, but not included: "d" in "World!" (position -2):
# eans start = -5 and end = -2-1 = -3 

b = "Hello, World!"
print(b[-5:-2])

orl


In [58]:
b = "Hello, World"
b[-1]  

'd'

In [59]:
# here start index is nothing means from beginning and end index is -1 hence -1-1 = -2 it will end at -2 
b = "Hello, World"
b[:-1] 

'Hello, Worl'

In [60]:
b = "Hello, World"
print(b[-1:])
print(b[:-1])
print(b[-5:])
print(b[-4:-2])


d
Hello, Worl
World
or


In [61]:
b = "Hello, World"
b[11:20]  # here 11 is alreday last index where 'd' is present. and if we give END value anything bigger then
# start then it will not be error

'd'

`Tip :  Always consider as [LB:UB] --LB index will be included and UB = UB-1 `
so 

starting point/index/bound = LB 

and ending point/inedx/bound   = UB-1



### Setp size in slicing

#We can also use index and slice notation to grab elements of a sequence by a specified step size (the default is 1). 
For instance we can use two colons in a row and then a number specifying the frequency to grab elements. For example:


### String - Upper()

Python has a set of built-in methods that you can use on strings.

The upper() method returns a string where all characters are in upper case. Symbols and Numbers are ignored.

> Syntax
string.upper()

> Parameter Values
No parameters

In [62]:
txt = "Hello @ my 1234 friends"
x = txt.upper()
print(x)

HELLO @ MY 1234 FRIENDS


### String - lower ()

The lower() method returns the string in lower case:

The lower() method returns a string where all characters are lower case.

 Symbols and Numbers are ignored.

In [63]:
txt = "Hello @ my 1234 friends"
x = txt.lower()
print(x)

hello @ my 1234 friends


### String replace()

The replace() method replaces a specified phrase with another specified phrase.

`Note: All occurrences of the specified phrase will be replaced, if nothing else is specified.`

Syntax

string.replace(oldvalue, newvalue, count)

Parameter Values

Parameter	Description

oldvalue	Required. The string to search for

newvalue	Required. The string to replace the old value with

count	Optional. A number specifying how many occurrences of the old value you want to replace. Default is all occurrences

In [64]:
# Signature: txt.replace(old, new, count=-1, /)
# Docstring:
# Return a copy with all occurrences of substring old replaced by new.
#
#   count
#     Maximum number of occurrences to replace. -1 (the default value) means replace all occurrences.
#
# If the optional argument count is given, only the first count occurrences are # replaced.
# Type:      builtin_function_or_method

txt = "I like bananas1 and more bananas2 and bananas3"

x1 = txt.replace("bananas", "apples")
x2 = txt.replace("bananas", "apples", 2)

print(x1)
print(x2)

I like apples1 and more apples2 and apples3
I like apples1 and more apples2 and bananas3


In [65]:
# Replace all occurrence of the word "one"

txt = "one one was a race horse, two two was one too."

x = txt.replace("one", "three")

print(x)

three three was a race horse, two two was three too.


In [66]:
# Replace the  first two occurrence of the word "one":

txt = "one one was a race horse, two two was one too."

x = txt.replace("one", "three", 2)
print(x)

three three was a race horse, two two was one too.


In [67]:
a = "Hello, World!"
print(a.replace("l", "J"))

HeJJo, WorJd!


### String split() 

`Definition and Usage
The split() method splits a string into a list.
You can specify the separator, default separator is any whitespace.
Note: When maxsplit is specified, the list will contain the specified number of elements plus one.
Syntax_
string.split(separator, maxsplit)_
Parameter Values
Parameter	Description
separator	Optional. Specifies the separator to use when splitting the string. By default any whitespace is a separator
maxsplit	Optional. Specifies how many splits to do. Default value is -1, which is "all occurrences`



In [68]:
# Split the string, using comma, followed by a space, as a separator:

txt = "hello, my name is Peter, I am 26 years old"

x = txt.split(", ") # here seprator is ,and one space 

print(x)

['hello', 'my name is Peter', 'I am 26 years old']


In [69]:
# The split() method splits the string into substrings if it finds instances of the separator

a = "Hello, World!"
print(a.split(",")) # returns ['Hello', ' World!']

['Hello', ' World!']


In [70]:
x

['hello', 'my name is Peter', 'I am 26 years old']

In [71]:
# Use a hash character as a separator:

txt = "apple#banana#cherry#orange"

x = txt.split("#")

print(x)

['apple', 'banana', 'cherry', 'orange']


In [72]:
# Use a separator which is not found in string

txt = "apple#banana#cherry#orange"

x = txt.split("##")

print(x)

['apple#banana#cherry#orange']


In [73]:
# dont give any separator

txt = "apple banana#cherry#orange"

x = txt.split() # default separator "one space" is applied 

print(x)

['apple', 'banana#cherry#orange']


In [74]:
# separator can not be empty. it will be ValueError: empty separator
txt = "apple  banana#cherry#orange"
x = txt.split("")

ValueError: empty separator

In [75]:
txt = "apple  banana#cherry#orange"
x = txt.split(" ")
print(len(x))

print()

print(x)
print()

count = 0 
for i in x:
    count+=1
    print(count, i, len(i), type(i))

3

['apple', '', 'banana#cherry#orange']

1 apple 5 <class 'str'>
2  0 <class 'str'>
3 banana#cherry#orange 20 <class 'str'>


In [76]:
# Split the string into a list with max 2 items:

txt = "apple#banana#cherry#orange"

# setting the maxsplit parameter to 1, will return a list with 2 elements!
x = txt.split("#", 1)

print(x)

['apple', 'banana#cherry#orange']


### String strip()

The strip() method removes any whitespace from the beginning or the end.

Whitespace is the space before and/or after the actual text, and very often you want to remove this space.

__Definition and Usage <br>__
The strip() method removes any leading (spaces at the beginning) and trailing (spaces at the end) characters (space is the default leading character to remove, __you can specifiy your own character as well__)

_Syntax_
string.strip(characters)

Parameter Values

Parameter	Description

characters	Optional. A set of characters to remove as leading/trailing characters

In [77]:
 # Signature: txt.strip(chars=None, /)
# Docstring:
# Return a copy of the string with leading and trailing whitespace removed.
# 
# If chars is given and not None, remove characters in chars instead.
# Type:      builtin_function_or_method

a = "        Hello, World!        "
print(a.strip())  # returns "Hello, World!"

Hello, World!


In [78]:
txt = "     banana     "

x = txt.strip()

print("of all fruits", x, "is my favorite")

of all fruits banana is my favorite


* __Not only whitespace you can remove any character spacied in .strip()__

Example :

In [79]:
# Remove the leading and trailing characters:
txt = ",,,,,rrttgg.....banana....rrr"
x = txt.strip(",.grt")
print(x)

banana


### String Concatenation

To concatenate, or combine, two strings you can use the + operator.

Merge variable a with variable b into variable c:




In [80]:
a = "Hello"
b = "World"
c = a + b
print(c)

HelloWorld


In [81]:
a = "Hello "
b = "World$"
c = a + b
print(c)

Hello World$


In [82]:
# To add a space between them, add a " ":
a = "Hello"
b = "World"
c = a + " " + b
print(c)

Hello World


In [83]:
# => INT+ int is possible 
a = 8
b = 3
print((a+b),type(a+b))

11 <class 'int'>


In [84]:
# => INT+ FLOAT is possible 
a = 5 
b = 1.5
print((a+b),type(a+b))

6.5 <class 'float'>


In [85]:
# int+complex also possbile 
a = 5
b = 6j
print((a+b),type(a+b))


(5+6j) <class 'complex'>


In [86]:
# int + String is ERROR 
a = 5
b = "John"
print(a + b)

TypeError: unsupported operand type(s) for +: 'int' and 'str'

In [87]:
age = 36
print("My name is John, I am " + age) # => int+String will give erro

TypeError: can only concatenate str (not "int") to str

# Format - Strings (format())


Learn more about String Formatting in String Formatting(https://www.w3schools.com/python/python_string_formatting.asp)


As we learned in the Python Variables chapter, we cannot combine strings and numbers like this:

`TypeError: can only concatenate str (not "int") to str`

In [88]:
age = 36
txt = "My name is John, I am " + age
print(txt)

TypeError: can only concatenate str (not "int") to str

But we can combine strings and numbers by using the `format()` method!

The format() method takes the passed arguments, formats them, and places them in the string where the placeholders {} are:

In [89]:
# Use the format() method to insert numbers into strings:

age = 36
txt = "My name is John, and I am {}"
print(txt.format(age))

My name is John, and I am 36


The format() method takes unlimited number of arguments, and are placed into the respective placeholders:

In [90]:
quantity = 3
itemno = 567
price = 49.95
myorder = "I want {} pieces of item {} for {} dollars."
print(myorder.format(quantity, itemno, price))

I want 3 pieces of item 567 for 49.95 dollars.


__You can use index numbers {0} to be sure the arguments are placed in the correct placeholders__

In [91]:
quantity = 3
itemno = 567
price = 49.95
myorder = "I want {2} pieces of item {0} for {1} dollars."
print(myorder.format(itemno, price, quantity))

I want 3 pieces of item 567 for 49.95 dollars.


### String Formatting

String formatting lets you inject items into a string rather than trying to chain items together using commas or string concatenation. As a quick comparison, consider:

    player = 'Thomas'
    points = 33
    
    'Last night, '+player+' scored '+str(points)+' points.'  # concatenation
    
    f'Last night, {player} scored {points} points.'          # string formatting


There are three ways to perform string formatting.
* The oldest method involves placeholders using the modulo `%` character.
* An improved technique uses the `.format()` string method.
* The newest method, introduced with Python 3.6, uses formatted string literals, called *f-strings*.

Since you will likely encounter all three versions in someone else's code, we describe each of them here.

In [92]:
player = 'Thomas'
points = 33

In [93]:
print('Last night, ' + player + ' scored ' + str(points) + ' points.')  # concatenation
print(f'Last night, {player} scored {points} points.')  # string formatting

Last night, Thomas scored 33 points.
Last night, Thomas scored 33 points.


## Formatting with placeholders (%) (%s, %r)
You can use <code>%s</code> to inject strings into your print statements. The modulo `%` is referred to as a "string formatting operator".


In [94]:
print("I'm going to inject %s here." % 'something')

I'm going to inject something here.


In [95]:
print("I'm going to inject %s here." % ('something'))

I'm going to inject something here.


You can pass multiple items by placing them inside a tuple after the `%` operator.

In [96]:
print("I'm going to inject %s text here, and %s text here." %('some','more'))

I'm going to inject some text here, and more text here.


You can also pass variable names:

In [97]:
x, y = 'some', 'more'
print("I'm going to inject %s text here, and %s text here."%(x,y))

I'm going to inject some text here, and more text here.


### Format conversion methods.
It should be noted that two methods <code>%s</code> and <code>%r</code> convert any python object to a string using two separate methods: `str()` and `repr()`. We will learn more about these functions later on in the course, but you should note that `%r` and `repr()` deliver the *string representation* of the object, including quotation marks and any escape characters.

In [98]:
print('He said his name was %s.' %'Fred')
print('He said his name was %r.' %'Fred')

He said his name was Fred.
He said his name was 'Fred'.


As another example, `\t` inserts a tab into a string.

In [99]:
print('I once caught a fish %s.' %'this \tbig')
print('I once caught a fish %r.' %'this \tbig')

I once caught a fish this 	big.
I once caught a fish 'this \tbig'.


HERE;

`%r` take eveyting as string hence it displayed  %'this \tbig' as  'this \tbig'

where as 
`%s` take  'this \tbig' and  apply the escape char also.



The `%s` operator converts whatever it sees into a string, including integers and floats. 

The `%d` operator converts numbers to integers first, without rounding. 

Note the difference below:

In [100]:
print('I wrote %s programs today.' %3.75)
print('I wrote %d programs today.' %3.75)   

I wrote 3.75 programs today.
I wrote 3 programs today.


### repr()

https://www.programiz.com/python-programming/methods/built-in/repr

lets Learn about the Python repr() function with the help of examples.

The repr() function returns a printable representation of the given object.

means it will return a string and  whatever you pased in repr() will be considered as String type (i.e. str)

 syntax of repr() is: 
 
repr(obj)

Parameters

The repr() function takes a single parameter:

obj - the object whose printable representation has to be returned

repr() Return Value <br>
The repr() function returns a printable representational string of the given object.


In [101]:
numbers = [1, 2, 3, 4, 5]

# create a printable representation of the list. 
printable_numbers = repr(numbers)
print(printable_numbers)
print(type(printable_numbers))

[1, 2, 3, 4, 5]
<class 'str'>


In [102]:
numbers = [1, 2, 3, 4, 5]
numbers

[1, 2, 3, 4, 5]

In [103]:
type(numbers)

list

In [104]:
printable_numbers

'[1, 2, 3, 4, 5]'

In [105]:
type(printable_numbers)

str

__How repr() works in Python?__

In [106]:
var = 'foo'
repr(var)

"'foo'"

Here, we assign a value 'foo' to var. Then, the repr() function returns "'foo'" which is 'foo' inside double-quotes. 

point to understand here is actaul value is not only `foo` but `'foo'`  --> that's the repr() does

### str() 

Convert the number 3.5 into an string

Definition and Usage

The str() function converts the specified value into a string.

Syntax

str(object, encoding=encoding, errors=errors)

str(object, encoding='utf-8', errors='strict')

Parameter Values

Parameter	Description

object - The object whose string representation is to be returned. If not provided, returns the empty string
        Any object. Specifies the object to convert into a string

encoding - Encoding of the given object. Defaults of UTF-8 when not provided.

errors - Specifies what to do if the decoding fails,
          Response when decoding fails. Defaults to 'strict'.


There are six types of errors:

* strict - default response which raises a UnicodeDecodeError exception on failure
* ignore - ignores the unencodable Unicode from the result
* replace - replaces the unencodable Unicode to a question mark
* xmlcharrefreplace - inserts XML character reference instead of unencodable Unicode
* backslashreplace - inserts a \uNNNN espace sequence instead of unencodable Unicode
* namereplace - inserts a \N{...} escape sequence instead of unencodable Unicode


_Return value from str()_

The str() method returns a string, which is considered an informal or nicely printable representation of the given object.
 
  
   

In [107]:
# Convert a string into an integer
x = int("12")
x

12

In [108]:
# Convert an integer into string
x = str(12)
x

'12'

__Convert to String__

If encoding and errors parameter isn't provided, str() internally calls the `__str__()` method of an object.

If it cannot find the `__str__()` method, it instead calls `repr(obj)`

In [109]:
result = str(10)
result

'10'

Try these commands on the Python console.

In [110]:
str('Adam')

'Adam'

In [111]:
str(b'Python!')

"b'Python!'"

### Padding and Precision of Floating Point Numbers
Floating point numbers use the format <code>%5.2f</code>. Here, <code>5</code> would be the minimum number of characters the string should contain; these may be padded with whitespace if the entire number does not have this many digits. Next to this, <code>.2f</code> stands for how many numbers to show past the decimal point. Let's see some examples:

Floating-point numbers use the format %a.bf. Here, a would be the minimum number of digits to be present in the string; these might be padded with white space if the whole number doesn’t have this many digits. Close to this, bf represents how many digits are to be displayed after the decimal point. 


‘%s’ is used to inject strings similarly 

‘%d’ for integers, 

‘%f’ for floating-point values, 

‘%b’ for binary format. 

For all formats, conversion methods visit the official documentation.

`NOTE : here in %5.2f 5 is minimum length of result inclusing "." not the maximum
        here in %10.2f 10 is minimum length of result inclusing "."not the maximum `

In [112]:
print('Integer numbers with placeholder:#%d' %3.75) 

Integer numbers with placeholder:#3


In [113]:
print('Floating point numbers with placeholder:#%5.2f' %(13.144))
# %5.2f  ==> entire string will be minimum 5 char including "."
# (i.e. maximum could be anyting as per number given )
# and after deciaml point 2 number will be shows from input 
# here like . is at . and mone
# after "." 14 as .2f mentioned ..so its .14 (3 char) out of 5 2 char are remaing so 
# 13.14 

Floating point numbers with placeholder:#13.14


In [114]:
print('Floating point numbers with placeholder:#%10.2f' %(13.144))
# %10.2f  ==> entire string will be minimum 10 char (i.e. maximum could be anyting as per number given )
# and after deciaml point 2 number will be shows from input 
# here like "." is alined with ","  and 
# after "." 14 as .2f mentioned ..so its .14 (3 char) out of 10-3 = 7 char are remaing so 7 spaces will be padded 
# SSSSSSS13.14  (here S = space)

Floating point numbers with placeholder:#     13.14


In [115]:
print('Floating point numbers with placeholder:#%1.0f' %(13.144))

Floating point numbers with placeholder:#13


In [116]:
print('Floating point numbers with placeholder:#%f' %(13.144))

Floating point numbers with placeholder:#13.144000


In [117]:
print('Floating point numbers with placeholder:#%9.5f' %(13.144))

Floating point numbers with placeholder:# 13.14400


In [118]:
print('Floating point numbers with placeholder:#%1.5F' % (13.144))

Floating point numbers with placeholder:#13.14400


In [119]:
print('Floating point numbers with placeholder:#%25.2f' %(13.144))

Floating point numbers with placeholder:#                    13.14


In [120]:
# minimum char will be 1 and after . zero digit hence only 13 will be result 
print("Floating point numbers with placeholder:1#%1.0f" % 13.144)

print("Floating point numbers with placeholder:2#%5.2f" % 13.149) # here its does rouding of 14 to 15 self 

print("Floating point numbers with placeholder:3#%1.5f" % 13.144)
# 13.14 and 5 spaces will be there 
print("Floating point numbers with placeholder:4#%10.2f" % 13.144) 

# 13.15 and 5 spaces will be there 
print("Floating point numbers with placeholder:5#%10.2f" % 13.149) 

# 13.57 and 5 spaces will be there 
print("Floating point numbers with placeholder:6#%10.2f" % 13.579) 

# No minium char so it take all. but after "." only 2 digit will be there 
# and in this case its getting round off so .57 will be .58 
print("Floating point numbers with placeholder:7#%.2f" % 13.579) 

print("20 white space Padded before", "with placeholder", "Floating point number:8#%25.2f" %13.144, sep="#")


Floating point numbers with placeholder:1#13
Floating point numbers with placeholder:2#13.15
Floating point numbers with placeholder:3#13.14400
Floating point numbers with placeholder:4#     13.14
Floating point numbers with placeholder:5#     13.15
Floating point numbers with placeholder:6#     13.58
Floating point numbers with placeholder:7#13.58
20 white space Padded before#with placeholder#Floating point number:8#                    13.14


In [121]:
variable = 12

string = "Variable as integer = %d \n\
Variable as float = %f" %(variable, variable)

print (string)


Variable as integer = 12 
Variable as float = 12.000000


In [122]:
print("binary of 5 is %b" % 5)

ValueError: unsupported format character 'b' (0x62) at index 16

For more information on string formatting with placeholders visit https://docs.python.org/3/library/stdtypes.html#old-string-formatting

### Multiple Formatting with placeholders (%)
Nothing prohibits using more than one conversion tool in the same print statement:

In [123]:
print('First: %s, Second:#%5.2f, Third: %r' % ('hi!', 3.1415, 'bye!'))

First: hi!, Second:# 3.14, Third: 'bye!'


## Formatting with the `.format()` method
A better way to format objects into your strings for print statements is with the string `.format()` method. The syntax is:

    'String here {} then also {}'.format('something1','something2')
    
For example:

In [124]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


In [125]:
'String here {} then also {}'.format('something1','something2')

'String here something1 then also something2'

In [126]:
print("I'm going to inject %s text here, and %s text here." %('some','more'))

I'm going to inject some text here, and more text here.


In [127]:
print("I'm going to inject {} text here, and {} text here.".format('some', 'more'))

I'm going to inject some text here, and more text here.


### String Formatting using  format()
To make sure a string will display as expected, we can format the result with the format() method.

The format() method allows you to format selected parts of a string.

Sometimes there are parts of a text that you do not control, maybe they come from a database, or user input?

To control such values, add placeholders (curly brackets {}) in the text, and run the values through the format() method:

In [128]:
price = 49
txt = "The price is {} dollars"
print(txt.format(price))

The price is 49 dollars


You can add parameters inside the curly brackets to specify how to convert the value

In [129]:
# Format the price to be displayed as a number with two decimals:

price = 49
txt = "The price is {:.2f} dollars"
print(txt.format(price))


The price is 49.00 dollars


__Multiple Values__

If you want to use more values, just add more values to the format() method:

print(txt.format(price, itemno, count))  And add more placeholders:

In [130]:
quantity = 3
itemno = 567
price = 49
myorder = "I want {} pieces of item number {} for {:.2f} dollars."
print(myorder.format(quantity, itemno, price))


I want 3 pieces of item number 567 for 49.00 dollars.


__Index Numbers__

You can use index numbers (a number inside the curly brackets {0}) to be sure the values are placed in the correct placeholders:

In [131]:
quantity = 3
itemno = 567
price = 49
myorder = "I want {0} pieces of item number {1} for {2:.2f} dollars."
print(myorder.format(quantity, itemno, price))

I want 3 pieces of item number 567 for 49.00 dollars.


Also, if you want to refer to the same value more than once, use the index number:

In [132]:
age = 36
name = "John"
txt = "His name is {1}. {1} is {0} years old."
print(txt.format(age, name))

His name is John. John is 36 years old.


__Named Indexes__

You can also use named indexes by entering a name inside the curly brackets {carname}, but then you must use names when you pass the parameter values txt.format(carname = "Ford"):

In [133]:
myorder = "I have a {carname}, it is a {model}."
print(myorder.format(carname="Ford", model="Mustang"))

I have a Ford, it is a Mustang.


__Definition and Usage__

The format() method formats the specified value(s) and insert them inside the string's placeholder.

The placeholder is defined using curly brackets: {}. Read more about the placeholders in the Placeholder section below.

The format() method returns the formatted string.


Syntax

string.format(value1, value2...)

Parameter Values

Parameter	Description

value1, value2...	Required. One or more values that should be formatted and inserted in the string.

The values are either a list of values separated by commas, a key=value list, or a combination of both.

The values can be of any data type.


__The Placeholders__
The placeholders can be identified using named indexes {price}, numbered indexes {0}, or even empty placeholders {}.


Formatting Types
Inside the placeholders you can add a formatting type to format the result:

:<		Left aligns the result (within the available space)

:>		Right aligns the result (within the available space)

:^		Center aligns the result (within the available space)

:=		Places the sign to the left most position

:+		Use a plus sign to indicate if the result is positive or negative

:-		Use a minus sign for negative values only

: 		Use a space to insert an extra space before positive numbers (and a minus sign before negative numbers)

:,		Use a comma as a thousand separator

:_		Use a underscore as a thousand separator

:b		Binary format

:c		Converts the value into the corresponding unicode character

:d		Decimal format

:e		Scientific format, with a lower case e

:E		Scientific format, with an upper case E

:f		Fix point number format

:F		Fix point number format, in uppercase format (show inf and nan as INF and NAN)

:g		General format

:G		General format (using a upper case E for scientific notations)

:o		Octal format

:x		Hex format, lower case

:X		Hex format, upper case

:n		Number format

:%		Percentage format




In [134]:
# To demonstrate, we insert the number 8 to set the available space for the value to 8 characters.
# Use "<" to left-align the value:
txt = "We have {:<8} chickens."
print(txt.format(49))

# Use ">" to right-align the value:
txt = "We have {:>8} chickens."
print(txt.format(49))

# Use "^" to center-align the value:

txt = "We have {:^8} chickens."
print(txt.format(49))

# Use "=" to place the plus/minus sign at the left most position:
txt = "The temperature is {:=8} degrees celsius."
print(txt.format(-5))

# Use "+" to always indicate if the number is positive or negative:
txt = "The temperature is between {:+} and {:+} degrees celsius."
print(txt.format(-3, 7))

# Use "-" to always indicate if the number is negative (positive numbers are displayed without any sign):
txt = "The temperature is between {:-} and {:-} degrees celsius."
print(txt.format(-3, 7))

# Use " " (a space) to;
# insert a space before positive numbers and
# insert a minus sign before negative numbers:
txt = "The temperature is between{: }and{: }degrees celsius."
print(txt.format(-3, 7))


# Use "," to add a comma as a thousand separator:
txt = "The universe is {:,} years old."
print(txt.format(13800000000))

#Use "_" to add a underscore character as a thousand separator:
txt = "The universe is {:_} years old."
print(txt.format(13800000000))

#Use "b" to convert the number into binary format:
txt = "The binary version of {0} is {0:b}"
print(txt.format(5))


# :c		Converts the value into the corresponding unicode character
txt = "unicode character for number {0} is {0:c} "
print(txt.format(256))

#Use "d" to convert a number, in this case a binary number, into decimal number format:
txt = "We have {:d} chickens."
print(txt.format(0b101))

#Use "e" to convert a number into scientific number format (with a lower-case e here e = 10.0^1):
txt = "We have {:e} chickens."
print(txt.format(5))

#Use "E" to convert a number into scientific number format (with an upper-case E):
txt = "We have {:E} chickens."
print(txt.format(5))

#Use "o" to convert the number into octal format:
txt = "The octal version of {0} is {0:o}"
print(txt.format(255))

#Use "x" to convert the number into Hex format:
txt = "The Hexadecimal version of {0} is {0:x}"
print(txt.format(255))

#Use "X" to convert the number into Hex format:
txt = "The Hexadecimal version of {0} is {0:X}"
print(txt.format(255))




We have 49       chickens.
We have       49 chickens.
We have    49    chickens.
The temperature is -      5 degrees celsius.
The temperature is between -3 and +7 degrees celsius.
The temperature is between -3 and 7 degrees celsius.
The temperature is between-3and 7degrees celsius.
The universe is 13,800,000,000 years old.
The universe is 13_800_000_000 years old.
The binary version of 5 is 101
unicode character for number 256 is Ā 
We have 5 chickens.
We have 5.000000e+00 chickens.
We have 5.000000E+00 chickens.
The octal version of 255 is 377
The Hexadecimal version of 255 is ff
The Hexadecimal version of 255 is FF


In [135]:
# Use "%" to convert the number into a percentage format:
txt = "You scored {:%}"
print(txt.format(0.25))
# Or, without any decimals:
txt = "You scored {:.0%}"
print(txt.format(0.25))


txt = "You scored {:%}"
print(txt.format(0.256))
# Or, without any decimals:
txt = "You scored {:.0%}"
print(txt.format(0.256))


You scored 25.000000%
You scored 25%
You scored 25.600000%
You scored 26%


In [136]:
# Use "f" to convert a number into a fixed point number, default with 6 decimals, 
# but use a period followed by a number to specify the number of decimals:
txt = "The price is {:.2f} dollars."
print(txt.format(45))

# without the ".2" inside the placeholder, this number will be displayed like this:
txt = "The price is {:f} dollars."
print(txt.format(45))

#Use "F" to convert a number into a fixed point number, but display inf and nan as INF and NAN:
x = float('inf')
txt = "The price is {:F} dollars."
print(txt.format(x))

#same example, but with a lower case f:
txt = "The price is {:f} dollars."
print(txt.format(x))

The price is 45.00 dollars.
The price is 45.000000 dollars.
The price is INF dollars.
The price is inf dollars.


In [137]:
x = float('inf')
print(x)
print(type(x))

inf
<class 'float'>


In [138]:
# Using different placeholder values:

# named indexes:
txt1 = "My name is {fname}, I'm {age}".format(fname="John", age=36)
print(txt1)

# numbered indexes:
txt2 = "My name is {0}, I'm {1}".format("John", 36)
print(txt2)

# empty placeholders:
txt3 = "My name is {}, I'm {}".format("John", 36)
print(txt3)

# The values are either a list of values separated by commas, a key=value list, or a combination of both.
txt4 = "My name is {key}, I'm {} lets {}".format(36, "play hocky", key="John")
print(txt4)

My name is John, I'm 36
My name is John, I'm 36
My name is John, I'm 36
My name is John, I'm 36 lets play hocky


In [139]:
x = float('inf')
print(x)
print(type(x))

inf
<class 'float'>


### The .format() method has several advantages over the %s placeholder method:

#### 1. Inserted objects can be called by index position:

In [140]:
print('The {2} {1} {0}'.format('fox', 'brown', 'quick'))

The quick brown fox


#### 2. Inserted objects can be assigned keywords:

In [141]:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1, b='Two', c=12.3))

First Object: 1, Second Object: Two, Third Object: 12.3


#### 3. Inserted objects can be reused, avoiding duplication:

In [142]:
print('A %s saved is a %s earned.' %('penny','penny'))
# vs.
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.
A penny saved is a penny earned.


### Alignment, padding and precision with `.format()`
Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more

In [143]:
print('{0:8} | {1:9}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:9}'.format('Apples', 3.))
print('{0:8} | {1:9}'.format('Oranges', 10))

Fruit    | Quantity 
Apples   |       3.0
Oranges  |        10


In [144]:
print('{0:8}|{1:9}'.format('Fruit', 'Quantity'))
print('{0:9}|{1:9}'.format('Apples', 3.))
print('{0:10}|{1:9}'.format('Oranges', 10))

Fruit   |Quantity 
Apples   |      3.0
Oranges   |       10


By default, `.format()` aligns text to the left, numbers to the right. You can pass an optional `<`,`^`, or `>` to set a left, center or right alignment:

In [145]:
print("{0:8},{1:8},{2:8}".format("Left", "Center", "Right"))
print("{0:<8}|{1:^8}|{2:>8}".format('Left', 'Center', 'Right'))
print("{0:<8}|{1:^8}|{2:>8}".format(11, 12, 13))

Left    ,Center  ,Right   
Left    | Center |   Right
11      |   12   |      13


`NOTE : By default, .format() aligns numbers to the right but here we are passing alignment option also `

You can precede the aligment operator with a padding character

In [146]:
print('Hi {0:=<8} | {1:-^9} | {2:.>8}'.format('Left', 'Center', 'Right'))
print('{0:=<8} | {1:-^9} | {2:.>8}'.format(11, 12, 13))

Hi Left==== | -Center-- | ...Right
11====== | ---12---- | ......13


In [147]:
print("{0:=<8}#|${1:=^8}|@{2:=>8}".format('Left', 'Center', 'Right'))
print("{0:=<8}#|${1:=^8}|@{2:=>8}".format(11, 12, 13))

Left====#|$=Center=|@===Right
11======#|$===12===|@======13


In [148]:
print("{0:=<8}#|${1:=^8}|@{2:=>8}".format('Left', 'Center', 'Right'))
# here we have asked to take 13 first , then 11 and last is 12 by putting different index values
print("{2:=<8}#|${0:=^8}|@{1:=>8}".format(11, 12, 13))

Left====#|$=Center=|@===Right
13======#|$===11===|@======12


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [149]:
print("Floating point numbers with placeholder: #%5.2f" % 13.149)
print("Floating point numbers with placeholder: #%1.0f" % 13.144)
print("Floating point numbers with placeholder: #%1.5f" % 13.144)
print("Floating point numbers with placeholder: #%10.2f" % 13.144) # 13 and 8 spaces will be there 
print("Floating point numbers with placeholder: #%10.2f" % 13.579) 
print("Floating point numbers with placeholder: #%.2f" % 13.579)  # no minium char so it take all 
print("20 white space Padded before", "with placeholder", "Floating point numbers:%25.2f" % 13.144, sep="#")

Floating point numbers with placeholder: #13.15
Floating point numbers with placeholder: #13
Floating point numbers with placeholder: #13.14400
Floating point numbers with placeholder: #     13.14
Floating point numbers with placeholder: #     13.58
Floating point numbers with placeholder: #13.58
20 white space Padded before#with placeholder#Floating point numbers:                    13.14


In [150]:
print("Floating point numbers with format: #{0:5.2f}".format(13.149))
print("Floating point numbers with format: #{0:1.0f}".format(13.144))
print("Floating point numbers with format: #{0:1.5f}".format(13.144))
print("Floating point numbers with format: #{0:10.2f}".format(13.144)) # 13 and 8 spaces will be there 
print("Floating point numbers with format: #{0:10.2f}".format(13.579) )
print("Floating point numbers with format: #{0:.2f}".format(13.579))  # no minium char so it take all 
print("20 white space Padded before", "with placeholder", "Floating point numbers:#%25.2f".format(13.144), sep="#")

Floating point numbers with format: #13.15
Floating point numbers with format: #13
Floating point numbers with format: #13.14400
Floating point numbers with format: #     13.14
Floating point numbers with format: #     13.58
Floating point numbers with format: #13.58
20 white space Padded before#with placeholder#Floating point numbers:#%25.2f


In [151]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Other way could be ;

In [152]:
print('This is my ten-character, two-decimal number:{a}'.format(a=13.579))
print('This is my ten-character, two-decimal number:{a:10.2f}'.format(a=13.579))

This is my ten-character, two-decimal number:13.579
This is my ten-character, two-decimal number:     13.58


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.

For more information on the string `.format()` method visit https://docs.python.org/3/library/string.html#formatstrings

In [153]:
# Add a placeholder where you want to display the price:
price = 49
txt = "The price is {:.2f} dollars"
print(txt.format(price))

The price is 49.00 dollars


In [154]:
# Insert the price inside the placeholder, the price should be in fixed point, two-decimal format:
txt = "For only {price:.2f} dollars!"
print(txt.format(price = 49))

For only 49.00 dollars!


In [155]:
print("For only {price:.2f} dollars!".format(price = 49))

For only 49.00 dollars!


## Formatted String Literals (f-strings)

Introduced in Python 3.6, f-strings offer several benefits over the older `.format()` string method described above. 

For one, you can bring outside variables immediately into to the string rather than pass them as arguments through `.format(var)`.

In [156]:
name = 'Fred'

print(f"He said his name is {name}.")
print(f"He said His name is {name}.", end="@")

He said his name is Fred.
He said His name is Fred.@

Pass `!r` to get the string representation:

In [157]:
print(f"He said his name is {name!r}")
print(f"\nHe said His name is {name!r}.")

He said his name is 'Fred'

He said His name is 'Fred'.


In [158]:
name ='sam'
age = 3

In [159]:
print(f'{name} is {age} years old.')

sam is 3 years old.


In [160]:
f'{name} is {age} years old.'

'sam is 3 years old.'

#### Float formatting follows `"result: {value:{width}.{precision}}"`

https://saralgyaan.com/posts/f-string-in-python-usage-guide/#How_to_add_decimal_places_in_a_float_in_Python_f-strings_(Formatting_a_float)?

Where with the `.format()` method you might see `{value:10.4f}`, with f-strings this can become `{value:{10}.{6}}`


In [161]:
num = 23.46254
print("My 10 character, four decimal number is with place holder:#%10.4f" % num)

My 10 character, four decimal number is with place holder:#   23.4625


In [162]:
num = 23.45678
print("My 10 character, four decimal number is with format():#{0:10.4f}".format(num))

My 10 character, four decimal number is with format():#   23.4568


In [163]:
num = 23.46254

#  result:{10}.{6}}
# Rresult ki Total length 10 char hogi
# only 6 digit Display honge excluding "."  so like that 6 digitsd + 1 "." so = total 7 char display honge
# now if width given is more than 7 then rest postion will be filled by white space
# {num:{a}.{b}} => when a > b  No. of Digit displayed = b+1(for .) AND No. of whitespace =  a-(b+1)
# {num:{a}.{b}} => when a <= b No. of digit displayed = b+1(for .) AND No. of whitespace =  0

# Note : Width => This maximum  Width. Full length of result can NOT go more that Width but can go less than Width

# here
# num:{10}.{6}} => when 10 > 6  No. of char to be displayed = 6+1(for .) = 7
#         AND No. of whitespace =  10-(6+1) = 10- 7 = 3
print(f"with f-string {num} num:(10).(6) is:#{num:{10}.{6}}")

with f-string 23.46254 num:(10).(6) is:#   23.4625


In [164]:
num = 23.46254
print(f"with f-string {num} num:(3).(5) is:#{num:{3}.{5}}")

with f-string 23.46254 num:(3).(5) is:#23.463


In [165]:
num = 123.46254
print(f"with f-string {num} num:(3).(5) is:#{num:{3}.{5}}")

with f-string 123.46254 num:(3).(5) is:#123.46


In [166]:
num = 9123.46254
print(f"with f-string {num} num:(3).(5) is:#{num:{3}.{5}}")

with f-string 9123.46254 num:(3).(5) is:#9123.5


In [167]:
num = 79123.46254
print(f"with f-string {num} num:(10).(6) is:#{num:{3}.{5}}")

with f-string 79123.46254 num:(10).(6) is:#7.9123e+04


In [168]:
print(f"with f-string(23.45678:(0).(4))  is:#{23.45678:{0}.{4}}")
print(f"with f-string(23.45678:(1).(4))  is:#{23.45678:{1}.{4}}")
print(f"with f-string(23.45678:(10).(4)) is:#{23.45678:{10}.{4}}")
print(f"with f-string(23.45678:(10).(5)) is:#{23.45678:{10}.{5}}")
print(f"with f-string(23.45678:(10).(6)) is:#{23.45678:{10}.{6}}")
print(f"with f-string(23.45678:(10).(7)) is:#{23.45678:{10}.{7}}")

# If char are more (i.e. 14 in this case ) than width (10)
print(f"with f-string(1234567.45678:(10).(8)) is:#{1234567.45678:{10}.{8}}")

# If char are less (i.e. 7 in this case ) than width(10)
print(f"with f-string(9.45678:(10).(8)) is:#{9.45678:{10}.{8}}")

with f-string(23.45678:(0).(4))  is:#23.46
with f-string(23.45678:(1).(4))  is:#23.46
with f-string(23.45678:(10).(4)) is:#     23.46
with f-string(23.45678:(10).(5)) is:#    23.457
with f-string(23.45678:(10).(6)) is:#   23.4568
with f-string(23.45678:(10).(7)) is:#  23.45678
with f-string(1234567.45678:(10).(8)) is:# 1234567.5
with f-string(9.45678:(10).(8)) is:#   9.45678


In [169]:
print(f"with f-string(123.456723:(8).(6))  is:#{123.456723:{8}.{6}}")
print(f"with f-string(123.4:(8).(6))  is:#{123.4:{8}.{6}}")

print(f"with f-string(123.456723:(8).(4))  is:#{123.456723:{8}.{4}}")
print(f"with f-string(123.456723:(8).(2))  is:#{123.456723:{8}.{2}}")
print(f"with f-string(123.456723:(8).(3))  is:#{123.456723:{8}.{3}}")

with f-string(123.456723:(8).(6))  is:# 123.457
with f-string(123.4:(8).(6))  is:#   123.4
with f-string(123.456723:(8).(4))  is:#   123.5
with f-string(123.456723:(8).(2))  is:# 1.2e+02
with f-string(123.456723:(8).(3))  is:#1.23e+02


In [170]:
result = 100/777
print("result=", result)
print()
print("The result was {}".format(result))
print("The result was {r}".format(r=result))
print("The result was {r:1.3f}".format(r=result))
print("The result was {r:10.3f}".format(r=result))
print("The result was {r:10.5f}".format(r=result))
print("The result was {r:1.5f}".format(r=result))
print("The result was {r:1.5f}".format(r=104.12345))
print("The result was {r:1.2f}".format(r=104.12345))

result= 0.1287001287001287

The result was 0.1287001287001287
The result was 0.1287001287001287
The result was 0.129
The result was      0.129
The result was    0.12870
The result was 0.12870
The result was 104.12345
The result was 104.12


Note :  **that with f-strings, *precision* refers to the total number of digits, not just those following the decimal.** 

This fits more closely with scientific notation and statistical analysis. Unfortunately, f-strings do not pad to the right of the decimal, even if precision allows it: (i.e. zero padding at right side doent happen in f-string)

In [171]:
num = 23.45
print("My 10 character, four decimal number is:#{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:#{num:{10}.{6}}")

My 10 character, four decimal number is:#   23.4500
My 10 character, four decimal number is:#     23.45


If this becomes important, you can always use `.format()` method syntax inside an f-string:

In [172]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:10.4f}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:   23.4500


For more info on formatted string literals visit https://docs.python.org/3/reference/lexical_analysis.html#f-strings

In [173]:
print(f"{'Left':=<8}@|${'center':=^8}|@{'Right':=>8}")
print(f"{'13':=<8}@|${'12':=^8}|@{'11':=>8}")
print(f"{13:=<8}@|${12:=^8}|@{11:=>8}")

Left====@|$=center=|@===Right
13======@|$===12===|@======11
13======@|$===12===|@======11


That is the basics of string formatting!

## Python - Escape Characters

Escape Character

To insert characters that are illegal in a string, use an escape character.

An escape character is a backslash `\` followed by the character you want to insert.

An example of an illegal character is a double quote inside a string that is surrounded by double quotes:

In [174]:
# You will get an error if you use double quotes inside a string that is surrounded by double quotes:
txt = "We are the so-called "Vikings" from the north."
print(txt)

SyntaxError: invalid syntax (<ipython-input-174-1305ada74cd2>, line 2)

to fix this problem, use the escape character `\"`:

In [175]:
txt = "We are the so-called \"Vikings\" from the north."
print(txt)

We are the so-called "Vikings" from the north.


__Other escape characters used in Python:__

> Code	Result	
`\'`	Single Quote	
`\\`	Backslash	
`\n`	New Line	
`\r`	Carriage Return	
`\t`	Tab	
`\b`	Backspace	
`\f`	Form Feed	
`\ooo`	Octal value	
`\xhh`	Hex value

In [176]:
# \'	Single Quote
txt2 = "This will insert one \' (Single Quote)."
print(txt2)
print()

# \\	Backslash
txt1 = "This will insert one \\ (backslash)."
print(txt1)
print()

# \t	Tab
txt3 = "This will insert one$\t@(tab)."
print(txt3)
print()

# \n	New Line
txt3 = "This will insert one$\n@(new line)."
print(txt3)
print()

# \b	Backspace
# This example erases one character (backspace):
txt5 = "This will insert one$\b@(backspace and delete $)."
print(txt5)
txt4 = "Hello \bWorld!"
print(txt4)

print()

# xhh	Hex value
# A backslash followed by an 'x' and a hex number represents a hex value:
txt6 = "x48x65x6cx6cx6f"
print(txt6)
txt7 = "\x48\x65\x6c\x6c\x6f"
print(txt7)
print()

# ooo	Octal value
# A backslash f1ollowed by three integers will result in a octal value:
txt = "\110\145\154\154\157"  # hello
print(txt)
txt = "\117\143\164\141\154"  # Octal
print(txt)
print()

# \r Carriage Return
txt = "Hello\rWorld!"
print(txt)

# \f	Form Feed

This will insert one ' (Single Quote).

This will insert one \ (backslash).

This will insert one$	@(tab).

This will insert one$
@(new line).

This will insert one$@(backspace and delete $).
Hello World!

x48x65x6cx6cx6f
Hello

Hello
Octal

World!


In [177]:
print("Hey!", "It\'s #\nalright.", sep='@', end='$')

Hey!@It's #
alright.$

# Python - String Methods

Python has a set of built-in methods that you can use on strings.

`Note: All string methods returns new values. They do not change the original string.`

https://www.w3schools.com/python/python_strings_methods.asp

`Method			Description
capitalize()	Converts the first character to upper case
casefold()		Converts string into lower case
center()		Returns a centered string
count()			Returns the number of times a specified value occurs in a string
encode()		Returns an encoded version of the string
endswith()		Returns true if the string ends with the specified value
expandtabs()	Sets the tab size of the string
find()			Searches the string for a specified value and returns the position of where it was found
format()		Formats specified values in a string
format_map()	Formats specified values in a string
index()			Searches the string for a specified value and returns the position of where it was found
isalnum()		Returns True if all characters in the string are alphanumeric
isalpha()		Returns True if all characters in the string are in the alphabet
isdecimal()		Returns True if all characters in the string are decimals
isdigit()		Returns True if all characters in the string are digits
isidentifier()	Returns True if the string is an identifier
islower()		Returns True if all characters in the string are lower case
isnumeric()		Returns True if all characters in the string are numeric
isprintable()	Returns True if all characters in the string are printable
isspace()		Returns True if all characters in the string are whitespaces
istitle()		Returns True if the string follows the rules of a title
isupper()		Returns True if all characters in the string are upper case
join()			Joins the elements of an iterable to the end of the string
ljust()			Returns a left justified version of the string
lower()			Converts a string into lower case
lstrip()		Returns a left trim version of the string
maketrans()		Returns a translation table to be used in translations
partition()		Returns a tuple where the string is parted into three parts
replace()		Returns a string where a specified value is replaced with a specified value
rfind()			Searches the string for a specified value and returns the last position of where it was found
rindex()		Searches the string for a specified value and returns the last position of where it was found
rjust()			Returns a right justified version of the string
rpartition()	Returns a tuple where the string is parted into three parts
rsplit()		Splits the string at the specified separator, and returns a list
rstrip()		Returns a right trim version of the string
split()			Splits the string at the specified separator, and returns a list
splitlines()	Splits the string at line breaks and returns a list
startswith()	Returns true if the string starts with the specified value
strip()			Returns a trimmed version of the string
swapcase()		Swaps cases, lower case becomes upper case and vice versa
title()			Converts the first character of each word to upper case
translate()		Returns a translated string
upper()			Converts a string into upper case
zfill()			Fills the string with a specified number of 0 values at the beginning`

## capitalize()
(https://www.w3schools.com/python/ref_string_capitalize.asp)

Python String capitalize() Method

Example
Upper case the first letter in this sentence:

txt = "hello, and welcome to my world."

x = txt.capitalize()

print (x)

Definition and Usage

The capitalize() method returns a string where the first character is upper case, and the rest is lower case.

Syntax

string.capitalize()

Parameter Values

No parameters

More Examples

Example

The first character is converted to upper case, and the rest are converted to lower case:

txt = "python is FUN!"

x = txt.capitalize()

print (x)

Example

See what happens if the first character is a number:

txt = "36 is my age."

x = txt.capitalize()

print (x)

In [178]:
# Upper case the first letter in this sentence:

txt = "hello, and welcome to my world."

x = txt.capitalize()

print (x)

Hello, and welcome to my world.


In [179]:
# The first character is converted to upper case, and the rest are converted to lower case:

txt = "python is FUN!"

x = txt.capitalize()

print (x)

Python is fun!


In [180]:
# See what happens if the first character is a number:

txt = "36 is my age."

x = txt.capitalize()

print (x)

36 is my age.


## casefold()

https://www.w3schools.com/python/ref_string_casefold.asp

Python String casefold() Method

Example

Make the string lower case:

txt = "Hello, And Welcome To My World!"

x = txt.casefold()

print(x)

Definition and Usage

The casefold() method returns a string where all the characters are lower case.

This method is similar to the lower() method, but the casefold() method is stronger, more aggressive, meaning that it will convert more characters into lower case, and will find more matches when comparing two strings and both are converted using the casefold() method.

Syntax

string.casefold()

Parameter Values

No parameters




In [181]:
# Make the string lower case:

txt = "Hello, And Welcome To My World!"

x = txt.casefold()

print(x)


txt2 = "Hello, And Welcome To My World!"

x2 = txt2.lower()

print(x2)

hello, and welcome to my world!
hello, and welcome to my world!


## center()

https://www.w3schools.com/python/ref_string_center.asp

`Example
Print the word "banana", taking up the space of 20 characters, with "banana" in the middle:
txt = "banana"
x = txt.center(20)
print(x)
Definition and Usage
The center() method will center align the string, using a specified character (space is default) as the fill character.
Syntax
string.center(length, character)
Parameter Values
Parameter	Description
length	Required. The length of the returned string
character	Optional. The character to fill the missing space on each side. Default is " " (space)
More Examples
Example
Using the letter "O" as the padding character:
txt = "banana"
x = txt.center(20, "O")
print(x)`



In [182]:
# Print the word "banana", taking up the space of 20 characters, with "banana" in the middle:
txt = "banana"
x = txt.center(20)
print(x)

       banana       


In [183]:
# Using the letter "O" as the padding character:
txt = "banana"
x = txt.center(20, "$")
print(x)

$$$$$$$banana$$$$$$$


## count() 
https://www.w3schools.com/python/ref_string_count.asp

`Example
Return the number of times the value "apple" appears in the string:
txt = "I love apples, apple are my favorite fruit"
x = txt.count("apple")
print(x)
Definition and Usage
The count() method returns the number of times a specified value appears in the string.
Syntax
string.count(value, start, end)
Parameter Values
Parameter	Description
value	Required. A String. The string to value to search for
start	Optional. An Integer. The position to start the search. Default is 0
end	Optional. An Integer. The position to end the search. Default is the end of the string
More Examples
Example
Search from position 10 to 24:
txt = "I love apples, apple are my favorite fruit"
x = txt.count("apple", 10, 24)
print(x)`


In [184]:
# Return the number of times the value "apple" appears in the string:
txt = "I love apples, apple are my favorite fruit"
x = txt.count("apple")  # we have word as apples NOT apple still it will get counted 
print(x)

2


In [185]:
# Search from position 10 to 24:
txt = "I love apples, apple are my favorite fruit"
x = txt.count("apple", 10, 24)
print(x)
print(len(txt))

1
42


## lower()

https://www.w3schools.com/python/ref_string_lower.asp

`Example
Lower case the string:
txt = "Hello my FRIENDS"
x = txt.lower()
print(x)
Definition and Usage
The lower() method returns a string where all characters are lower case.
Symbols and Numbers are ignored.
Syntax
string.lower()
Parameter Values
No parameters`





In [186]:
# Lower case the string:
txt = "Hello my FRIENDS"
x = txt.lower()
print(x)

hello my friends


## upper()

https://www.w3schools.com/python/ref_string_upper.asp

`Example
Upper case the string:
txt = "Hello my friends"
x = txt.upper()
print(x)
Definition and Usage
The upper() method returns a string where all characters are in upper case.
Symbols and Numbers are ignored.
Syntax
string.upper()
Parameter Values
No parameters`




In [187]:
# Upper case the string:
txt = "Hello my friends"
x = txt.upper()
print(x)

HELLO MY FRIENDS
